libcaf_core/test/settings.cpp

```diff
@@ -0,0 +1,162 @@
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#define CAF_SUITE settings
+
+#include "caf/settings.hpp"
+
+#include "caf/test/dsl.hpp"
+
+#include <string>
+
+#include "caf/atom.hpp"
+#include "caf/none.hpp"
+#include "caf/optional.hpp"
+
+using namespace caf;
+
+namespace {
+
+// TODO: switch to std::operator""s when switching to C++14
+std::string operator"" _s(const char* str, size_t size) {
+  return std::string(str, size);
+}
+
+struct fixture {
+  settings x;
+
+  void fill() {
+    x["hello"] = "world";
+    x["one"].as_dictionary()["two"].as_dictionary()["three"] = 4;
+    auto& logger = x["logger"].as_dictionary();
+    logger["component-blacklist"] = make_config_value_list(atom("caf"));
+    logger["console"] = atom("none");
+    logger["console-format"] = "%m";
+    logger["console-verbosity"] = atom("trace");
+    logger["file-format"] = "%r %c %p %a %t %C %M %F:%L %m%n";
+    logger["inline-output"] = false;
+    auto& middleman = x["middleman"].as_dictionary();
+    middleman["app-identifiers"] = make_config_value_list("generic-caf-app");
+    middleman["enable-automatic-connections"] = false;
+    middleman["heartbeat-interval"] = 0;
+    middleman["max-consecutive-reads"] = 50;
+    middleman["workers"] = 3;
+    auto& stream = x["stream"].as_dictionary();
+    stream["credit-round-interval"] = timespan{10000000}; // 10ms;
+    stream["desired-batch-complexity"] = timespan{50000}; // 50us;
+    stream["max-batch-delay"] = timespan{5000000};        // 5ms;
+  }
+};
+
+const config_value& unpack(const settings& x, string_view key) {
+  auto i = x.find(key);
+  if (i == x.end())
+    CAF_FAIL("key not found in dictionary: " << key);
+  return i->second;
+}
+
+template <class... Ts>
+const config_value& unpack(const settings& x, string_view key,
+                           const char* next_key, Ts... keys) {
+  auto i = x.find(key);
+  if (i == x.end())
+    CAF_FAIL("key not found in dictionary: " << key);
+  if (!holds_alternative<settings>(i->second))
+    CAF_FAIL("value is not a dictionary: " << key);
+  return unpack(get<settings>(i->second), {next_key, strlen(next_key)},
+                keys...);
+}
+
+} // namespace
+
+CAF_TEST_FIXTURE_SCOPE(settings_tests, fixture)
+
+CAF_TEST(put) {
+  put(x, "foo", "bar");
+  put(x, "logger.console", atom("none"));
+  put(x, "one.two.three", "four");
+  CAF_CHECK_EQUAL(x.size(), 3u);
+  CAF_CHECK(x.contains("foo"));
+  CAF_CHECK(x.contains("logger"));
+  CAF_CHECK(x.contains("one"));
+  CAF_CHECK_EQUAL(unpack(x, "foo"), "bar"_s);
+  CAF_CHECK_EQUAL(unpack(x, "logger", "console"), atom("none"));
+  CAF_CHECK_EQUAL(unpack(x, "one", "two", "three"), "four"_s);
+  put(x, "logger.console", atom("trace"));
+  CAF_CHECK_EQUAL(unpack(x, "logger", "console"), atom("trace"));
+}
+
+CAF_TEST(put missing) {
```

Comment:
You are not checking that an already contained value isn't added again! Maybe try to add some contained value with a different key and check that the size stays the same and the old key + value is still contained afterwards.

Reply:
The last `put_missing` checks that, does it not?

```c++
  put_missing(x, "logger.console", atom("trace"));
  CAF_CHECK_EQUAL(unpack(x, "logger", "console"), atom("none"));
```

Reply:
True. I misread `atom("trace")` for `atom("none")`.
I'm still confused by the description of the function regarding this test:
```
/// Converts `value` to a `config_value` and assigns it to `key` if `value` is
/// currently missing in `xs`.
```
Shouldn't you try to add the same value with a different key? The description doesn't say anything about handling same keys.

Reply:
Documentation fail. Will fix.